src/function/gds/K_Core_Decomposition.cpp

```diff
@@ -0,0 +1,324 @@
+                if (dontCheckOffset || curDenseFrontier->isActive(offset)) {
+                    curSmallest = std::min(curSmallest,
+                        curVertexValues.getData(tableID)[offset].load(std::memory_order_relaxed));
+                }
+            }
+        }
+        uint64_t current = curSmallestDegree.load(std::memory_order_relaxed);
+        curSmallestDegree.compare_exchange_strong(current, curSmallest, std::memory_order_relaxed);
+    }
+
+    void setAllActive() {
+        for (const auto& [tableID, curNumNodes] : numNodesMap) {
+            curDenseFrontier->pinTableID(tableID);
+            for (uint64_t offset = 0; offset < curNumNodes; ++offset) {
+                common::nodeID_t nodeID;
+                nodeID.tableID = tableID;
+                nodeID.offset = offset;
+                curDenseFrontier->setActive(nodeID);
+            }
+        }
+    }
+
+    uint64_t getSmallestDegree() { return curSmallestDegree.load(std::memory_order_relaxed); }
+
+    void setUpdateFlag() { updateDegreeFlag = true; }
+
+    bool isNodeActive(common::nodeID_t nodeID) const {
+        return curDenseFrontier->isActive(nodeID.offset);
+    }
+
+    uint64_t getVertexValue(common::offset_t offset) {
+        return vertexValues[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    bool updateDegreeFlag = false;
+    std::atomic<uint64_t> curSmallestDegree{UINT64_MAX};
+    common::table_id_map_t<common::offset_t> numNodesMap;
+    std::atomic<uint64_t>* vertexValues = nullptr;
+    ObjectArraysMap<std::atomic<uint64_t>> curVertexValues;
+};
+
+struct KCoreInitEdgeCompute : public EdgeCompute {
+    KCoreFrontierPair* frontierPair;
+
+    explicit KCoreInitEdgeCompute(KCoreFrontierPair* frontierPair) : frontierPair{frontierPair} {}
+
+    std::vector<common::nodeID_t> edgeCompute(common::nodeID_t boundNodeID,
+        graph::NbrScanState::Chunk& chunk, bool) override {
+        std::vector<common::nodeID_t> result;
+        uint64_t nbrAmount = 0;
+        chunk.forEach([&](auto /* nbrNodeID */, auto) { nbrAmount++; });
+        result.push_back(boundNodeID);
+        frontierPair->addToVertexDegree(boundNodeID, nbrAmount);
+        return result;
+    }
+
+    std::unique_ptr<EdgeCompute> copy() override {
+        return std::make_unique<KCoreInitEdgeCompute>(frontierPair);
+    }
+};
+
+struct KCoreEdgeCompute : public EdgeCompute {
+    KCoreFrontierPair* frontierPair;
+
+    explicit KCoreEdgeCompute(KCoreFrontierPair* frontierPair) : frontierPair{frontierPair} {}
+
+    std::vector<common::nodeID_t> edgeCompute(common::nodeID_t boundNodeID,
+        graph::NbrScanState::Chunk& chunk, bool) override {
+        std::vector<common::nodeID_t> result;
+        uint64_t vertexDegree = frontierPair->getVertexValue(boundNodeID.offset);
+        uint64_t smallestDegree = frontierPair->getSmallestDegree();
+        if (frontierPair->isNodeActive(boundNodeID) && (vertexDegree == smallestDegree)) {
```

Comment:
If you are able to enter edgeCompute then boundNodeID is guaranteed to be active right?